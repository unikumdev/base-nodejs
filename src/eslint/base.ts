import { join } from 'path'

import type { Linter } from 'eslint'

export const getBase = ({
  isReact,
  pathFileTSConfig,
}: {
  readonly isReact?: boolean
  readonly pathFileTSConfig: string
}) => {
  const rulesReact: Linter.RulesRecord = isReact
    ? {
        'react/button-has-type': 'error',
        'react/jsx-newline': ['error', { prevent: false }],
        'react/jsx-sort-default-props': 'error',
        'react/jsx-sort-props': 'error',
        'react/no-children-prop': 'off',
        'react/no-unused-prop-types': 'error',
        'react/no-unused-state': 'error',
        'react/prefer-stateless-function': [
          'error',
          { ignorePureComponents: true },
        ],
        'react/self-closing-comp': 'error',
        'react/sort-comp': 'error',
        'react/sort-prop-types': 'error',
      }
    : {}

  return {
    env: {
      es2021: true,
      node: true,
    },
    extends: [
      'eslint-config-airbnb-base',
      isReact && 'plugin:react/recommended',
      isReact
        ? 'eslint-config-airbnb-typescript'
        : 'eslint-config-airbnb-typescript/base',
      'plugin:sort/recommended',
      'plugin:@typescript-eslint/recommended',
      'plugin:@typescript-eslint/recommended-requiring-type-checking',
      'plugin:prettier/recommended',
    ].filter(Boolean),
    parser: '@typescript-eslint/parser',
    parserOptions: {
      extraFileExtensions: ['.mjs'],
      project: [pathFileTSConfig, '**/*.js', '**/*.ts'],
    },
    plugins: ['@typescript-eslint', 'prefer-arrow', 'prettier', 'sort'],
    root: true,
    rules: {
      ...rulesReact,
      '@typescript-eslint/explicit-module-boundary-types': 'off',
      '@typescript-eslint/no-explicit-any': 'off',
      '@typescript-eslint/no-unsafe-assignment': 'off',
      '@typescript-eslint/no-unsafe-call': 'off',
      '@typescript-eslint/no-unsafe-member-access': 'off',
      '@typescript-eslint/no-unsafe-return': 'off',
      '@typescript-eslint/no-unused-vars': 'off',
      '@typescript-eslint/no-var-requires': 'off',
      '@typescript-eslint/unbound-method': 'off',
      'arrow-body-style': ['error', 'as-needed'],
      'block-spacing': 'error',
      'comma-dangle': [
        'error',
        {
          arrays: 'only-multiline',
          exports: 'only-multiline',
          functions: 'only-multiline',
          imports: 'only-multiline',
          objects: 'only-multiline',
        },
      ],
      'eol-last': 'error',
      'import/order': [
        'error',
        {
          alphabetize: {
            caseInsensitive: false,
            /* sort in ascending order. Options: ['ignore', 'asc', 'desc'] */
            order: 'asc',
          },
          groups: [
            'builtin',
            'external',
            'index',
            'internal',
            'object',
            'type',
            ['sibling', 'parent'],
          ],
          'newlines-between': 'always',
          pathGroups: [
            {
              group: 'internal',
              pattern: '@this/**/*',
              position: 'after',
            },
          ],
          pathGroupsExcludedImportTypes: ['internal'],
        },
      ],
      'import/prefer-default-export': 'off',
      'max-len': [
        'error',
        {
          code: 80,
          comments: 120,
          ignoreRegExpLiterals: true,
          ignoreStrings: true,
          ignoreTemplateLiterals: true,
          ignoreUrls: true,
          tabWidth: 2,
        },
      ],
      'no-multi-spaces': 'error',
      'no-multiple-empty-lines': [
        'error',
        {
          max: 1,
          maxEOF: 0,
        },
      ],
      'no-plusplus': 'off',
      'no-restricted-syntax': 'off',
      'no-underscore-dangle': 'off',
      'no-unused-vars': 'error',
      'object-curly-spacing': ['error', 'always'],
      'padding-line-between-statements': [
        'error',
        {
          blankLine: 'always',
          next: '*',
          prev: [
            'block-like',
            'block',
            'class',
            'const',
            'directive',
            'for',
            'function',
            'iife',
            'let',
            'switch',
            'try',
            'var',
            'while',
          ],
        },
        {
          blankLine: 'any',
          next: ['const', 'let', 'var'],
          prev: ['const', 'let', 'var'],
        },
        { blankLine: 'always', next: '*', prev: ['case', 'default'] },
      ],
      'prefer-arrow-callback': ['error', { allowNamedFunctions: true }],
      'prefer-arrow/prefer-arrow-functions': [
        'error',
        {
          classPropertiesAllowed: true,
          disallowPrototype: true,
          singleReturnOnly: true,
        },
      ],
      'prettier/prettier': 'error',
      quotes: ['error', 'single'],
      semi: ['error', 'never'],
      'sort/imports': 'off',
      'sort/type-properties': 'error',
    } as typeof rulesReact,
    settings: {
      react: {
        version: 'detect',
      },
    },
  }
}

/* istanbul ignore next */
export const getBaseESLint = <
  T1 extends Parameters<typeof getBase>[0] & {
    readonly pathDirRoot: string
  },
>(
  options: T1,
) => {
  /* istanbul ignore next */
  const baseConfig = getBase(options)

  // eslint-disable-next-line import/no-dynamic-require, global-require
  const packageJSON = require(join(options.pathDirRoot, 'package.json')) as {
    readonly devDependencies?: {
      readonly [x: string]: string
    }
  }

  if (packageJSON.devDependencies?.jest) {
    return {
      ...baseConfig,
      env: {
        ...baseConfig.env,
        jest: true,
        'jest/globals': true,
      },

      extends: [...baseConfig.extends, 'plugin:jest/recommended'],
    }
  }

  /* istanbul ignore next */
  return baseConfig
}
