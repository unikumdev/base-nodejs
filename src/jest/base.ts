import { merge } from 'merge-anything'

import { TSConfig } from '@this/src/modules/tsconfig'

import type { Config } from '@jest/types'
import type * as typeSWC from '@swc/core'
import type * as typeTypeScript from 'typescript'

export type { Config }

const defaults = {
  configs: {
    '@swc/jest': {
      jsc: {
        externalHelpers: true,
        parser: {
          dynamicImport: true,
          syntax: 'typescript',
          tsx: false,
        },
        target: 'es2021',
        transform: {
          react: {
            runtime: 'automatic',
          },
        },
      },
      module: {
        type: 'es6',
      },
      sourceMaps: true,
    },
  },

  moduleNameMapper: {
    // @TODO remove this when SWC has fixed https://github.com/swc-project/swc/issues/2753
    // right now the jsc.paths does nothing
    '@this/(.*)': '<rootDir>',
    '\\.(css|less|scss|sss|styl|sass)$':
      '<rootDir>/node_modules/identity-obj-proxy',
    // ESM needs a `.js` file extension
    '^(\\.{1,2}/.*)\\.js$': '$1',
  },
}

const getSWCPaths = <
  T1 extends {
    readonly paths?: typeTypeScript.MapLike<string[]>
    readonly pathDirRoot: string
  },
>(
  options: T1,
) => {
  if (options.paths) {
    Object.keys(options.paths).forEach((k) => {
      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion, no-param-reassign
      options.paths![k] = options.paths![k].reduce((acc, x) => {
        if (x.startsWith('./')) {
          acc.push(x.replace('./', `${options.pathDirRoot}/`))
        } else {
          acc.push(x)
        }

        return acc
      }, [] as string[])
    })
  }

  return options.paths
}

export const getBase = ({
  pathDirRoot,
  pathFileTSConfig,
}: {
  readonly pathDirRoot?: string
  readonly pathFileTSConfig?: string
} = {}): Config.InitialOptions => {
  const shouldReadTSConfig = Boolean(pathFileTSConfig && pathDirRoot)
  const contenFileTSConfig: {
    readonly compilerOptions?: typeTypeScript.CompilerOptions
  } = shouldReadTSConfig
    ? // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      TSConfig.readTSConfig({ pathFile: pathFileTSConfig! })
    : {}

  return {
    coveragePathIgnorePatterns: ['<rootDir>/build/', '<rootDir>/dist/'],
    coverageReporters: ['html-spa', 'lcov'],
    extensionsToTreatAsEsm: ['.ts', '.tsx'],
    moduleFileExtensions: [
      'cjs',
      'js',
      'json',
      'jsx',
      'mjs',
      'node',
      'ts',
      'tsx',
      'wasm',
    ],
    moduleNameMapper: {
      ...defaults.moduleNameMapper,
    },
    rootDir: pathDirRoot,
    testEnvironment: 'node',
    testMatch: ['**/__tests__/**/*.spec.[t]s?(x)'],
    transform: {
      '^.+\\.(t|j)sx?$': [
        '@swc/jest',
        merge(defaults.configs['@swc/jest'], {
          jsc: {
            baseUrl: contenFileTSConfig.compilerOptions?.baseUrl || undefined,
            parser: {
              decorators:
                contenFileTSConfig.compilerOptions?.experimentalDecorators ||
                false,
              tsx:
                /* istanbul ignore next */
                (contenFileTSConfig.compilerOptions?.jsx && true) ||
                defaults.configs['@swc/jest'].jsc.parser.tsx,
            } as typeSWC.TsParserConfig,
            paths:
              pathDirRoot && contenFileTSConfig.compilerOptions?.paths
                ? getSWCPaths({
                    pathDirRoot,
                    paths: contenFileTSConfig.compilerOptions.paths,
                  })
                : undefined,
            target:
              contenFileTSConfig.compilerOptions?.target ||
              defaults.configs['@swc/jest'].jsc?.target,
            transform: {
              decoratorMetadata:
                contenFileTSConfig.compilerOptions?.emitDecoratorMetadata ||
                false,
              legacyDecorator:
                contenFileTSConfig.compilerOptions?.experimentalDecorators ||
                false,
            },
          },
          // needed for snapthots, user may not override this
          sourceMaps: true,
        } as Partial<typeSWC.Config> as any),
      ],
    },
  }
}
