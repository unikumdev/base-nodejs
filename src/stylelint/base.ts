import type * as stylelint from 'stylelint'

export const getBase = (): stylelint.Config => ({
  defaultSeverity: 'error',
  extends: [
    'stylelint-config-sass-guidelines',
    'stylelint-config-prettier-scss',
  ],
  overrides: [],
  plugins: [],
  rules: {
    'declaration-empty-line-before': [
      'always',
      {
        ignore: [
          'after-comment',
          'after-declaration',
          'first-nested',
          'inside-single-line-block',
        ],
      },
    ],
    'no-eol-whitespace': true,
    'no-missing-end-of-source-newline': true,
    'scss/at-extend-no-missing-placeholder': true,
  },
})
