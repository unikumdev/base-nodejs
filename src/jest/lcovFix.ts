#!/usr/bin/env node

/**
 * https://github.com/kulshekhar/ts-jest/issues/542
 */
import * as path from 'path'

import {
  access,
  constants as constantsFS,
  ensureDir,
  readFile,
  writeFile,
} from 'fs-extra'
import { Argv } from 'yargs'
import { hideBin } from 'yargs/helpers'
import yargs from 'yargs/yargs'

export interface IOptionsCLI {
  readonly pathFile?: string
  readonly pathFileOutput?: string
}

export class LCOVFix {
  static defaults = {
    encoding: 'utf-8',
  }

  static tasks = {
    async checks({ pathFile, pathFileOutput }: IOptionsCLI = {}) {
      if (!pathFile) {
        // eslint-disable-next-line no-console
        console.error(
          '\x1b[31m%s\x1b[0m',
          'argument --path-file[string] is missing',
        )
        throw new Error('program exit(1)')
      }

      const pathFileResolved = path.resolve(pathFile)

      await access(pathFileResolved, constantsFS.R_OK)

      return {
        pathFile: pathFileResolved,
        pathFileOutput: pathFileOutput
          ? path.resolve(pathFileOutput)
          : pathFileResolved,
      }
    },
  }

  static async fix({
    pathFile,
    pathFileOutput,
  }: {
    readonly pathFile: string
    readonly pathFileOutput: string
  }) {
    const pathPrefix = process.cwd()
    const rawFile = await readFile(pathFile, 'utf8')
    const rebuiltPaths = rawFile
      .split('\n')
      .map((singleLine) => {
        if (singleLine.startsWith('SF:') && !singleLine.startsWith('SF:/')) {
          return singleLine.replace('SF:', `SF:${pathPrefix}/`)
        }

        return singleLine
      })
      .join('\n')

    // eslint-disable-next-line no-console
    console.info(rebuiltPaths)

    await ensureDir(path.dirname(pathFileOutput))

    return writeFile(pathFileOutput, rebuiltPaths, 'utf8')
  }

  static async execute(options: IOptionsCLI) {
    const { pathFile, pathFileOutput } = await LCOVFix.tasks.checks(options)

    return LCOVFix.fix({ pathFile, pathFileOutput })
  }
}

/* istanbul ignore next */
if (process.env.NODE_ENV !== 'test') {
  const { argv } = yargs(hideBin(process.argv)) as Argv<
    Parameters<typeof LCOVFix['execute']>[0]
  >
  const { pathFile, pathFileOutput } = argv

  LCOVFix.execute({ pathFile, pathFileOutput }).catch((err) => {
    // eslint-disable-next-line no-console
    console.error(err)
    process.exit(1)
  })
}
