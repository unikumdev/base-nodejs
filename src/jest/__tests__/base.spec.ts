/* eslint-disable max-len */
import { join } from 'path'

import { file } from 'tmp-promise'

import * as theModule from '@this/src/jest/base'

describe('jest base', () => {
  const defaults = {
    paths: {
      dirs: {
        root: join(__dirname, '../../..'),
      },
    },
  }
  const filterConfigResults = (x: ReturnType<typeof theModule['getBase']>) => {
    // contains local paths
    // eslint-disable-next-line no-param-reassign
    x.rootDir = undefined

    return x
  }

  describe('getBase', () => {
    describe('options', () => {
      it.concurrent('absent', () => {
        expect(filterConfigResults(theModule.getBase())).toMatchSnapshot()
      })

      describe('pathFileTSConfig', () => {
        it.concurrent('provided (fake file)', async () => {
          const { cleanup, path: pathFileTSConfig } = await file()

          expect(
            filterConfigResults(
              theModule.getBase({
                pathDirRoot: defaults.paths.dirs.root,
                pathFileTSConfig,
              }),
            ),
          ).toMatchSnapshot()
          return cleanup()
        })

        it.concurrent('real file', () => {
          const result = theModule.getBase({
            pathDirRoot: defaults.paths.dirs.root,
            pathFileTSConfig: join(defaults.paths.dirs.root, 'tsconfig.json'),
          })

          // paths for the real file will be translated with local path
          const keyFirstTransform = Object.keys(result.transform || {})[0]

          expect(
            ((result.transform || {})[keyFirstTransform][1] as any).jsc.paths,
          ).toBeDefined()
          ;((result.transform || {})[keyFirstTransform][1] as any).jsc.paths =
            undefined

          expect(filterConfigResults(result)).toMatchSnapshot()
        })
      })
    })
  })
})
