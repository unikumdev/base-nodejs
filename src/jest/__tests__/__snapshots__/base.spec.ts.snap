// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
Object {
  "coveragePathIgnorePatterns": Array [
    "<rootDir>/build/",
    "<rootDir>/dist/",
  ],
  "coverageReporters": Array [
    "html-spa",
    "lcov",
  ],
  "extensionsToTreatAsEsm": Array [
    ".ts",
    ".tsx",
  ],
  "moduleFileExtensions": Array [
    "cjs",
    "js",
    "json",
    "jsx",
    "mjs",
    "node",
    "ts",
    "tsx",
    "wasm",
  ],
  "moduleNameMapper": Object {
    "@this/(.*)": "<rootDir>",
    "\\\\.(css|less|scss|sss|styl|sass)$": "<rootDir>/node_modules/identity-obj-proxy",
    "^(\\\\.{1,2}/.*)\\\\.js$": "$1",
  },
  "rootDir": undefined,
  "testEnvironment": "node",
  "testMatch": Array [
    "**/__tests__/**/*.spec.[t]s?(x)",
  ],
  "transform": Object {
    "^.+\\\\.(t|j)sx?$": Array [
      "@swc/jest",
      Object {
        "jsc": Object {
          "baseUrl": undefined,
          "externalHelpers": true,
          "parser": Object {
            "decorators": false,
            "dynamicImport": true,
            "syntax": "typescript",
            "tsx": false,
          },
          "paths": undefined,
          "target": "es2021",
          "transform": Object {
            "decoratorMetadata": false,
            "legacyDecorator": false,
            "react": Object {
              "runtime": "automatic",
            },
          },
        },
        "module": Object {
          "type": "es6",
        },
        "sourceMaps": true,
      },
    ],
  },
}
`;

exports[` 2`] = `
Object {
  "coveragePathIgnorePatterns": Array [
    "<rootDir>/build/",
    "<rootDir>/dist/",
  ],
  "coverageReporters": Array [
    "html-spa",
    "lcov",
  ],
  "extensionsToTreatAsEsm": Array [
    ".ts",
    ".tsx",
  ],
  "moduleFileExtensions": Array [
    "cjs",
    "js",
    "json",
    "jsx",
    "mjs",
    "node",
    "ts",
    "tsx",
    "wasm",
  ],
  "moduleNameMapper": Object {
    "@this/(.*)": "<rootDir>",
    "\\\\.(css|less|scss|sss|styl|sass)$": "<rootDir>/node_modules/identity-obj-proxy",
    "^(\\\\.{1,2}/.*)\\\\.js$": "$1",
  },
  "rootDir": undefined,
  "testEnvironment": "node",
  "testMatch": Array [
    "**/__tests__/**/*.spec.[t]s?(x)",
  ],
  "transform": Object {
    "^.+\\\\.(t|j)sx?$": Array [
      "@swc/jest",
      Object {
        "jsc": Object {
          "baseUrl": ".",
          "externalHelpers": true,
          "parser": Object {
            "decorators": false,
            "dynamicImport": true,
            "syntax": "typescript",
            "tsx": true,
          },
          "paths": undefined,
          "target": "es2021",
          "transform": Object {
            "decoratorMetadata": false,
            "legacyDecorator": false,
            "react": Object {
              "runtime": "automatic",
            },
          },
        },
        "module": Object {
          "type": "es6",
        },
        "sourceMaps": true,
      },
    ],
  },
}
`;

exports[`jest base getBase options pathFileTSConfig provided (fake file) 1`] = `
Object {
  "coveragePathIgnorePatterns": Array [
    "<rootDir>/build/",
    "<rootDir>/dist/",
  ],
  "coverageReporters": Array [
    "html-spa",
    "lcov",
  ],
  "extensionsToTreatAsEsm": Array [
    ".ts",
    ".tsx",
  ],
  "moduleFileExtensions": Array [
    "cjs",
    "js",
    "json",
    "jsx",
    "mjs",
    "node",
    "ts",
    "tsx",
    "wasm",
  ],
  "moduleNameMapper": Object {
    "@this/(.*)": "<rootDir>",
    "\\\\.(css|less|scss|sss|styl|sass)$": "<rootDir>/node_modules/identity-obj-proxy",
    "^(\\\\.{1,2}/.*)\\\\.js$": "$1",
  },
  "rootDir": undefined,
  "testEnvironment": "node",
  "testMatch": Array [
    "**/__tests__/**/*.spec.[t]s?(x)",
  ],
  "transform": Object {
    "^.+\\\\.(t|j)sx?$": Array [
      "@swc/jest",
      Object {
        "jsc": Object {
          "baseUrl": undefined,
          "externalHelpers": true,
          "parser": Object {
            "decorators": false,
            "dynamicImport": true,
            "syntax": "typescript",
            "tsx": false,
          },
          "paths": undefined,
          "target": "es2021",
          "transform": Object {
            "decoratorMetadata": false,
            "legacyDecorator": false,
            "react": Object {
              "runtime": "automatic",
            },
          },
        },
        "module": Object {
          "type": "es6",
        },
        "sourceMaps": true,
      },
    ],
  },
}
`;
